server-protocol.md

```diff
@@ -92,6 +85,27 @@ following keys (and ignores all others):
   future version of the protocol requires a rate-limiting CAPTCHA ticket or
   other authorization record, the server can send `error` (explaining the
   requirement) if it does not see this ticket arrive before the `bind`.
+* `permission-required`: a set of available authentication methods,
+  proof of work challenges etc. The client needs to "solve" one of
+  them in order to get access to the service.
+
+The client should examine the `permissions-required` methods (if
+present) and select one to use (see also "Permission to Use the
+Server" below). If the client doesn't support any permissions methods
+or has already chosen not to use any then it _may_ send `bind`
+immediately upon connecting. Older clients that do not know about
+`permsision-required` will always send the `bind` immediately (without
+waiting for the `welcome` message). In that case, the server MAY grant usage permission 
+regardless. Otherwise, it will send a "permission denied" error message and close the connection.
+
+No matter when it is sent, the `bind` messages specifies the AppID and
+side (in keys `appid` and `side`, respectively) that all subsequent
+messages will be scoped to.  While technically each message could be
+independent (with its own `appid` and `side`), I thought it would be
+less confusing to use exactly one WebSocket per logical wormhole
+connection.  The `bind` may also include a `submit-permission` key
```

Comment:
We might consider delivering `submit-permission` in a distinct message, rather than bundling into the `bind` message. This could provide more room for a multi-round permission mechanism in the future.

Reply:
IIRC that's how we'd originally discussed it (and how I wrote that up originally) but I believe we converged on this to avoid an additional round-trip (right @piegamesde ?) although I suppose that doesn't help if one _wants_ an additional round-trip for some reason.

If the client sends `submit-permission` and then `bind` right away, we avoid the round-trip for current use-cases .. but then how would a future protocol add in a reply to the `submit-permission` if we don't already wait?

(I can think of SCRAM which needs an additional round-trip .. any other examples?)

Reply:
I think we'd pipeline the `submit-permission` + `bind` for the non-multi-round options, including `hashcash`. So the rule would be that the option type must be enough to indicate whether the client can send `submit-permission` right away, or if it needs to exchange more messages first. New modes can trigger the latter.

as for an example.. maybe something involving an external service, like reCAPTCHA or something, where the mailbox server needs to do something moderately expensive first (like talk to the service and initiate a CAPTCHA, receiving a URL to give to the client). We wouldn't want to do that unless the client was going to use it. So the sequence would be like:


| client/server | message                         | server/captcha-service |
| ----          | ----                            | ----                   |
| C -> S        | connects                        |                        |
| S -> C        | welcome([hashcash, captcha])    |                        |
| C -> S        | choose-permission(captcha)      |                        |
|               | start CAPTCHA                   | S -> captcha service   |
|               | URL                             | captcha service -> S   |
| S -> C        | captcha-url(URL)                |                        |
| C             | completes captcha               |                        |
| C -> S        | submit-permission(captcha: URL) |                        |
|               | check(URL)                      | S -> captcha service   |
|               | ok                              | captcha service -> S   |
| S -> C        | captcha-good                    |                        |
| C -> S        | bind                            |                        |


If the captcha failed, the mailbox server might send `captcha-bad` instead of `captcha-good` and then allow the client to try a new captcha, without dropping the connection entirely.

Reply:
What would be some possible applications for multi-roundtrip authentification?

Regardless, I think that the current proposal leaves enough room for adding this in a backwards-compatible way:

- The server announces that it supports that permission in its welcome message
- The client responds with a custom new message type to do all the intermediate steps
- The details can be figured out later, you get the idea.

Are there any other benefits of having `submit-permission` in a separate  message?

Reply:
Just tidyness.. `bind` is about limiting the connection to a certain scope, and was (at least nominally) not mandatory, if we decided to include its parameters in every subsequent message instead (multiplexing several requests over a single connection). Extending it to also handle permission granting feels ever so slightly messier than introducing a new message that's sent just before the `bind` and which doesn't require a response.

Call me -0 on extending `bind` to include `permission` and +0 on adding a separate `submit-permission` message.

Reply:
I agree that using the `bind` message for this is a case of "things growing historically" and that the name does not match the intended usage.

Do you think multiplexing multiple WS connections is still a thing that might happen one day? Because I read the documentation on this                             as "we could have but it isn't worth it". In that case, would one even need to authenticate multiple times?

Reply:
Gaining permission feels like "definitely a once-per-websocket connection thing" whereas I believe what @warner is saying is that the `bind` is (conceptually at least) an optional thing. So I would expect in a multiplexed situation, a client would "gain permission" once and then use any number of other operations.

I think that the most-granular thing we'd want to have "permissions" on would be at the "AppID" level, not e.g. different permissions for different nameplates/mailboxes -- that seems rife for problems with tracking? (I haven't tried to quantify that part, even in my head ;)

If we do add `submit-permission` as a separate message, for now that would look like:

 - send `submit-permission`
 - (immediately) also send `bind`

The server would process the `submit-permission` and either end the connection then or allow further processing (i.e. `bind` etc). (That is, if it is demanding `hashcash` at the moment, it's a protocol error to send anything except `submit-permission` next). This _may_ also make the state-machines cleaner? (See last statement).

This would keep things slightly tidier (separating the concerns of "permission-granting" from "binding") and allow an easier upgrade path for clients in the future. That is, if we added a `submit-permission` kind that needed more round-trips those clients understanding that permission would _not_ immediately send the `bind` if they select such a permission scheme (and instead wait for whatever the reply / more-round-trips message turns out to be). It's true they could instead just do the other messages (whatever they are) and finish with a `bind` (including the final `submit-permission`) but ....

....since `bind` is (at least conceptually) optional, a separate `submit-permission` messages does seem somewhat cleaner -- and maintains the benefit of "no more round-trips" for the immediate protocol changes while still providing a clean path towards more round-trips in the future. It also leaves the path open for multiplexing / "doing things without a `bind` at all" separate from the whole permissions concept.

So I think I'm slightly positive on "separate message" as well.

I _do_ have plans to implement this in the Python client and server in the next week or two, so we could hold off on merging this PR until we gain a little experience with at least one implementation and ensure we like one or the other...?

Reply:
Yes, I won't merge until all comments are resolved in the PR.

If you do it as a separate message, I'd suggest the following points to take care of:

- The `submit-permissions` is per WebSocket connection.
- Although the logical ordering requirements on the messages are little, I want a specified total order. This will make the implementations a lot easier.
- We can safely ignore multi-roundtrip authentication methods for now.